### PR TITLE
[proof] Further consolidation of the regular declaration path

### PR DESCRIPTION
**Pull request by ejgallego**
Miscellanous cleanings and refactoring in preparation for other PRs; most important change is the porting of a few users to the consolidated API in `DeclareDef`, as well as centralizing the preparation of constants there.

Future PRs will remove the remaining uses of low-level `Declare.declare_constant`. See note below on the meta-coq overlay.

Overlays:
- https://github.com/mattam82/Coq-Equations/pull/268
- https://github.com/coq-community/paramcoq/pull/50
- https://github.com/LPCIC/coq-elpi/pull/126
- https://github.com/MetaCoq/metacoq/pull/390

### CODE REVIEW

File: vernac/comAssumption.ml [hunk 1]
```diff
@@ -203,8 +203,11 @@ let context_insection sigma ~poly ctx =
           else Monomorphic_entry Univ.ContextSet.empty
         in
         let entry = Declare.definition_entry ~univs ~types:t b in
+        (* XXX Fixme: Use DeclareDef.prepare_definition *)
```
Comment (ejgallego) [inline vernac/comAssumption.ml]: This is already fixed in a later PR.

**Comment by ejgallego** [comment]
This is IMO ready for review; I will do a pass myself on the commits to check there are not silly typos as I wrote them a two weeks ago.

The metacoq overlay is interesting, cc @mattam82 @SkySkimmer .

`metacoq` failed because it did something like this on `Quote Recursively Definition foo := (fst, snd).`:

- first, it interpretes `(fst, snd)` in the usual way, implicits will get turned into unsolved evars;
- then, it proceeds to quote, and to use the evar_map resulting from the interpretation
- it calls `declare` on the quoted term, which after this PR will call `Pretyping.check_evars_are_solved` 
- it fails, as the `evar_map` we are using is stale w.r.t. term we want to save [the quoted one]

So indeed what I did is to reset the evar_map, but IMO this would be a good opportunity to add some docs about the preconditions we assume on the incoming evar_map at the prepare stage.

@mattam82 , this PR adds some docs to program mode, maybe you would like to do a pass on them.

**Comment by SkySkimmer** [comment]
>it fails, as the evar_map we are using is stale w.r.t. term we want to save

WDYM stale?

File: plugins/funind/gen_principle.ml [hunk 2]
```diff
@@ -290,12 +290,11 @@ let generate_functional_principle (evd: Evd.evar_map ref)
...
+  let entry = DeclareDef.ClosedDef.of_proof_entry ~uctx entry in
   let _ : Names.GlobRef.t = DeclareDef.declare_definition
-      ~name:new_princ_name ~hook_data
+      ~name:new_princ_name ~hook
       ~scope:(DeclareDef.Global Declare.ImportDefaultBehavior)
       ~kind:Decls.(IsProof Theorem)
-      ~ubind:UnivNames.empty_binders
       ~impargs:[]
       entry in
   ()
```
Comment (Matafou) [inline plugins/funind/gen_principle.ml]: Not sure to understand exactly the implication of this. Is it really the same?

Comment (ejgallego) [reply plugins/funind/gen_principle.ml]: It should, binders are now taken from the `evar_map` which seems the right pattern across the codebase.

**Comment by ejgallego** [comment]
> WDYM stale?

Indeed "stale" is likely the wrong term; it is more related to garbage-collecting `evars`.

MetaCoq does interpret a term which leaves a few unsolved evars, then proceeds to call `finalize` with a term unrelated to the `sigma` from the interpretation, `check_evars` then complains about the unsolved evars.

File: tactics/proof_global.ml [hunk 3]
```diff
@@ -18,9 +18,9 @@ module NamedDecl = Context.Named.Declaration
 
 type proof_object =
   { name : Names.Id.t
+  (* [name] only used in the STM *)
   ; entries : Evd.side_effects Declare.proof_entry list
   ; uctx: UState.t
```
Comment (ejgallego) [inline tactics/proof_global.ml]: Note that after this the proof object is basically a `uctx` and the `entr{y,ies}` , that is one of the goals here, and puts them in the same foot that what `DeclareDef.prepare` returns. In fact, if not for the name we could unify with `ClosedDef.t` already.

**Comment by SkySkimmer** [comment]
>check_evars then complains about the unsolved evars.

`check_evars` is actually a different function ;)
and maybe it's the function we should use instead of `check_evars_are_solved`

**Comment by ejgallego** [comment]
Bench (https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/895/console) shows no change.

> `check_evars` is actually a different function ;)
> and maybe it's the function we should use instead of `check_evars_are_solved`

Oh indeed sorry; I guess we should do the change in a different PR? Until metacoq arrived all the codepaths were indeed using the latter form, I'm unsure of the implications of this, TC resolution is far from my expertise.

**Comment by ejgallego** [comment]
IMHO this can go ahead; anyone willing to assign?

The `metacoq` patch is indeed interesting but IMHO the current solution is OK, and we can work out together finer approaches.

File: vernac/declareDef.mli [hunk 4]
```diff
@@ -39,23 +39,39 @@ module Hook : sig
...
+  -> impargs:Impargs.manual_implicits
+  -> uctx:UState.t
+  -> Evd.side_effects Declare.proof_entry
+  -> GlobRef.t
 
+(** Declares a non-interactive constant; [body] and [types] will be
+   normalized w.r.t. the passed [evar_map] [sigma]. Universes should
+   be handled properly, including minimization and restriction. Note
+   that [sigma] is checked for unresolved evars, thus you should be
+   careful not to submit open terms or evar maps with stale,
```
Comment (SkySkimmer) [inline vernac/declareDef.mli]: This could be read as meaning that it will anomaly on open terms, could you clarify that it's a regular error?

**Review by SkySkimmer — APPROVED** [review]

File: vernac/declareDef.ml [hunk 5]
```diff
@@ -65,10 +60,17 @@ let declare_definition ~name ~scope ~kind ?hook_data ~ubind ~impargs ce =
...
+      Hook.call ~hook { Hook.S.uctx; obls; scope; dref }
   end;
   dref
 
+let declare_definition ~name ~scope ~kind ?hook_data ~ubind ~impargs ce =
+  let fix_exn = Declare.Internal.get_fix_exn ce in
+  try declare_definition ~name ~scope ~kind ?hook_data ~ubind ~impargs ce
+  with exn ->
+    let exn = Exninfo.capture exn in
+    Exninfo.iraise (fix_exn exn)
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/declareDef.ml]: Inertia of the previous code, I'll amend.

File: vernac/lemmas.ml [hunk 6]
```diff
@@ -241,37 +223,39 @@ end = struct
...
-    List.map_i (declare_mutdef ~info ~uctx mutpe) 0 info.Info.thms
+    let pe = match info.Info.compute_guard with
+    | [] ->
+      (* Not a recursive statement *)
+      const
+    | possible_indexes ->
+      (* Try all combinations... not optimal *)
+      let env = Global.env() in
+      Declare.Internal.map_entry_body const
+        ~f:(guess_decreasing env possible_indexes)
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/lemmas.ml]: Trying to get closer to the common code / single path; I dunno.

File: vernac/declareDef.ml [hunk 7]
```diff
@@ -157,14 +157,37 @@ let check_definition_evars ~allow_evars sigma =
...
+  let allow_evars = true in
+  check_definition_evars ~allow_evars sigma;
+  let sigma, (body, types) = Evarutil.finalize ~abort_on_undefined_evars:(not allow_evars)
+      sigma (fun nf -> nf body, Option.map nf types)
+  in
+  let univs = Evd.check_univ_decl ~poly sigma udecl in
+  let ce = definition_entry ?opaque ?inline ?types ~univs body in
+  let env = Global.env () in
+  let (c,ctx), sideff = Future.force ce.Declare.proof_entry_body in
+  assert(Safe_typing.empty_private_constants = sideff.Evd.seff_private);
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/declareDef.ml]: I have fixed that in some commit; I'm not sure in which PR it is but already on the tree so we don't have to worry.

File: vernac/declareDef.ml [hunk 8]
```diff
@@ -153,23 +153,17 @@ let declare_assumption ?fix_exn ~name ~scope ~hook ~impargs ~uctx pe =
...
-  let env = Global.env () in
-  if not allow_evars then Pretyping.check_evars_are_solved ~program_mode:false env sigma
-
 let prepare_definition ?opaque ?inline ~poly ~udecl ~types ~body sigma =
-  let allow_evars = false in
-  check_definition_evars ~allow_evars sigma;
-  let sigma, (body, types) = Evarutil.finalize ~abort_on_undefined_evars:(not allow_evars)
+  let env = Global.env () in
+  Pretyping.check_evars_are_solved ~program_mode:false env sigma;
+  let sigma, (body, types) = Evarutil.finalize ~abort_on_undefined_evars:true
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/declareDef.ml]: Yup, why not, on the other hand it seems like a bit of useful documentation.

File: vernac/declareDef.mli (hunk 4 above)
Comment (ejgallego) [reply vernac/declareDef.mli]: Not sure what change you would like to see here; actually this doesn't always fails on open terms right? There is a call to TC resolution, tho with the failure flag on.

Comment (ejgallego) [reply vernac/declareDef.mli]: Please suggest an improved comment.

File: vernac/declareDef.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply vernac/declareDef.ml]: I think we want to keep it easy to grep for `abort_on_undefined_evars` and have it mean "not always true"

[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/declareDef.ml]: Good point, amended then!

File: vernac/declareDef.mli (hunk 4 above)
Comment (SkySkimmer) [reply vernac/declareDef.mli]: >Not sure what change you would like to see here

OK, let's move on

**Pull request merged**